### PR TITLE
get Max Date from promQL

### PR DESCRIPTION
**Pull request by msvinaykumar**
generateRecommendations API functioning without the specification of interval_end_time
It should be able to take the latest available timestamp in Prometheus
 finding the latest usage timestamp of a container

### CODE REVIEW

File: src/main/java/com/autotune/common/data/dataSourceQueries/DataSourceQueries.java [hunk 1]
```diff
@@ -17,7 +17,8 @@ public enum PromQLQuery {
...
-        MEMORY_REQUEST("%s by(container,namespace) (kube_pod_container_resource_requests{container!='', container!='POD', pod!='', namespace!='', namespace!~'kube-.*|openshift|openshift-.*', resource='memory', unit='byte',namespace=\"%s\",container=\"%s\"} * on(pod, namespace) group_left max by (container, pod, namespace) (kube_pod_status_phase{phase='Running'}))");
+        MEMORY_REQUEST("%s by(container,namespace) (kube_pod_container_resource_requests{container!='', container!='POD', pod!='', namespace!='', namespace!~'kube-.*|openshift|openshift-.*', resource='memory', unit='byte',namespace=\"%s\",container=\"%s\"} * on(pod, namespace) group_left max by (container, pod, namespace) (kube_pod_status_phase{phase='Running'}))"),
+        MAX_DATE("max(container_cpu_usage_seconds_total{container=\"%s\",namespace=\"%s\"} > 0)");
```
Comment (kusumachalasani) [inline src/main/java/com/autotune/common/data/dataSourceQueries/DataSourceQueries.java:21]: @msvinaykumar timestamp(max(container_cpu_usage_seconds_total{container=<>, namespace=<>} > 0)) can give the timestamp of that data directly.

Comment (msvinaykumar) [reply src/main/java/com/autotune/common/data/dataSourceQueries/DataSourceQueries.java:21]: ok

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 2]
```diff
@@ -200,6 +200,27 @@ public String validate() {
...
+    public String validate_local() {            //TODO Instead of relying on the 'local=true' check everywhere, aim to avoid this complexity by introducing a higher-level abstraction in the code.
+
+        String validationFailureMsg = "";
+        // Check if experiment_name is provided
+        if (experimentName == null || experimentName.isEmpty()) {
+            validationFailureMsg += AnalyzerErrorConstants.APIErrors.UpdateRecommendationsAPI.EXPERIMENT_NAME_MANDATORY + ", ";
+        }
+
+        // Check if interval_end_time is provided
+        if (intervalEndTimeStr != null) {
```
Comment (khansaad) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:212]: This is a redundant check, `isValidDate()` is already checking the `null` case.

Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:212]: I removed and tested without passing interval_end_time and it gives me this error 

"message": "Given timestamp - \" null \" is not a valid timestamp format",

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 3]
```diff
@@ -210,14 +231,18 @@ public KruizeObject prepareRecommendations(int calCount) {
...
+            interval_end_time = Utils.DateUtils.getTimeStampFrom(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT,
+                    intervalEndTimeStr);
+            setInterval_end_time(interval_end_time);
+        }
         KruizeObject kruizeObject = createKruizeObject();
         if (!kruizeObject.getValidation_data().isSuccess())
             return kruizeObject;
-        // continue to generate recommendation when kruizeObject is successfully created
         setKruizeObject(kruizeObject);
+        mainKruizeExperimentMAP.put(kruizeObject.getExperimentName(), kruizeObject);
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:244]: This Line should not affect ROS worklow it just update the map which is used to save recommendations

**Review by dinogun — APPROVED** [review]
Approving as this PoC code does not affect ROS workflow

**Pull request merged**